doc/design/to_run_stmt.md

```diff
@@ -0,0 +1,178 @@
+# To Run Statement
+
+## The Problem
+
+SQLFlow extends the SQL syntax and can describe an end-to-end machine learning pipeline. Data transformation is an important part in the entire pipeline. Currently we have the following two options for data transformation:
+
+- SQL.
+- COLUMN clause for per data instance transform. The transform logic is saved into the model.
+
+But that's not enough. It's a common case that the transformation is not per data instance which COLUMN clause is not suitable for. And we may need write very complex SQL or even cannot use SQL to express the transfomration logic. Let's take `extract_features` from [tsfresh package](https://tsfresh.readthedocs.io/en/latest/api/tsfresh.feature_extraction.html#module-tsfresh.feature_extraction.extraction) for example. It will preprocess the time series data, calculate statistical values using various analysis functions and parameters, and then generate tens or hunderds of features. [Convert the time series data of sliding window to rows](https://github.com/sql-machine-learning/sqlflow/issues/2238) is also a typical scenario.
+
+We can express the data transform logic using Python, leverage many mature python packages and encapsulate it into a reusable function. And then we propose to add the `TO RUN` clause in SQLFlow to call it.
+
+## Syntax Extension
+
+```SQL
+SELECT * FROM source_table
+TO RUN a_data_scientist/maxcompute_functions:1.0/my_func.data_preprocessor.a_python_func
+WITH param_a = value_a,
+     param_b = value_b
+INTO result_table
+```
+
+The SQLFlow statement above will call the python function `a_python_func` from the file `/run/my_func/data_preprocessor.py` inside the docker image `a_data_scientist/maxcompute_functions:1.0`. The attributes in the `WITH` clause will be passed to the python function as parameters.
+
+## Challenges
+
+- TO RUN can execute functions in Kubernetes and MaxCompute.
+- TO RUN can execute functions in both a single process and a distributed cluster.
+- TO RUN can execute functions built upon various data computing frameworks such as [Dask](https://github.com/dask/dask), [Mars](https://github.com/mars-project/mars), [Ray](https://github.com/ray-project/ray) and so on.
+
+## Design
+
+### Fix the challenges
+
+The semantics of `TO RUN a_data_scientist/maxcompute_functions:1.0/my_func.data_preprocessor.a_python_func` is that `a_python_func` will be executed in a docker container for Kubernetes or a PyODPS node for MaxCompute. The implementation of `a_python_func` is fully customized by users. Because the execution environments for `a_python_func` are different between Kubernetes and MaxCompute, there are some differences considering these three challenges above.
```

Comment:
This section is considering the execution environment/platform of TO RUN.  

In a container on Kubernetes and in a PyODPS node on MaxCompute are just two examples.  Not a complete list of environments.

Reply:
Good suggestion, I'll add a section `Execution Platforms`. 
And we will focus on the implementation details in Kubernetes and MaxCompute at this doc.